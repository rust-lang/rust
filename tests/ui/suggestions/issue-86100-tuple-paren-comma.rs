// Tests that a suggestion is issued for type mismatch errors when a
// 1-tuple is expected and a parenthesized expression of non-tuple
// type is supplied.

fn foo<T>(_t: (T,)) {}
struct S { _s: (String,) }

fn main() {
    let _x: (i32,) = (5);
    //~^ ERROR: mismatched types [E0308]
    //~| HELP: use a trailing comma to create a tuple with one element

    foo((Some(3)));
    //~^ ERROR: mismatched types [E0308]
    //~| HELP: use a trailing comma to create a tuple with one element

    let _s = S { _s: ("abc".to_string()) };
    //~^ ERROR: mismatched types [E0308]
    //~| HELP: use a trailing comma to create a tuple with one element

    // Do not issue the suggestion if the found type is already a tuple.
    let t = (1, 2);
    let _x: (i32,) = (t);
    //~^ ERROR: mismatched types [E0308]
}
