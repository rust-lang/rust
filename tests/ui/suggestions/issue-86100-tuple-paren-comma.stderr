error[E0308]: mismatched types
  --> $DIR/issue-86100-tuple-paren-comma.rs:9:22
   |
LL |     let _x: (i32,) = (5);
   |             ------   ^^^ expected `(i32,)`, found integer
   |             |
   |             expected due to this
   |
   = note: expected tuple `(i32,)`
               found type `{integer}`
help: use a trailing comma to create a tuple with one element
   |
LL |     let _x: (i32,) = (5,);
   |                        +

error[E0308]: mismatched types
  --> $DIR/issue-86100-tuple-paren-comma.rs:13:9
   |
LL |     foo((Some(3)));
   |     --- ^^^^^^^^^ expected `(_,)`, found `Option<{integer}>`
   |     |
   |     arguments to this function are incorrect
   |
   = note: expected tuple `(_,)`
               found enum `Option<{integer}>`
note: function defined here
  --> $DIR/issue-86100-tuple-paren-comma.rs:5:4
   |
LL | fn foo<T>(_t: (T,)) {}
   |    ^^^    --------
help: use a trailing comma to create a tuple with one element
   |
LL |     foo((Some(3),));
   |                 +

error[E0308]: mismatched types
  --> $DIR/issue-86100-tuple-paren-comma.rs:17:22
   |
LL |     let _s = S { _s: ("abc".to_string()) };
   |                      ^^^^^^^^^^^^^^^^^^^ expected `(String,)`, found `String`
   |
   = note: expected tuple `(String,)`
             found struct `String`
help: use a trailing comma to create a tuple with one element
   |
LL |     let _s = S { _s: ("abc".to_string(),) };
   |                                        +

error[E0308]: mismatched types
  --> $DIR/issue-86100-tuple-paren-comma.rs:23:22
   |
LL |     let _x: (i32,) = (t);
   |             ------   ^^^ expected a tuple with 1 element, found one with 2 elements
   |             |
   |             expected due to this
   |
   = note: expected tuple `(i32,)`
              found tuple `({integer}, {integer})`

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0308`.
